File: node-express/src/hook.js
import pino from 'pino';

const logger = pino({ name: 'custom-hook' });

// Mirrors the CustomHook from the Spring variant: before/after/error/finallyAfter
// on every flag evaluation so you can see the lifecycle in the logs.
export class CustomHook {
  before(hookContext, _hints) {
    logger.info({ flagKey: hookContext.flagKey }, 'Before hook');
    return undefined;
  }

  after(hookContext, details, _hints) {
    logger.info(
      { flagKey: hookContext.flagKey, reason: details.reason, variant: details.variant },
      'After hook',
    );
  }

  error(hookContext, error, _hints) {
    logger.error({ flagKey: hookContext.flagKey, err: error }, 'Error hook');
  }

  finallyAfter(hookContext, details, _hints) {
    logger.info(
      { flagKey: hookContext.flagKey, reason: details?.reason },
      'Finally after hook',
    );
  }
}

File: python-fastapi/app/hook.py
"""Custom OpenFeature hook that logs every lifecycle stage of a flag evaluation."""

from __future__ import annotations

import logging
from typing import Any

from openfeature.evaluation_context import EvaluationContext
from openfeature.flag_evaluation import FlagEvaluationDetails
from openfeature.hook import Hook, HookContext

LOG = logging.getLogger("app.hook")


class CustomHook(Hook):
    """Mirrors the Spring Boot CustomHook — logs before/after/error/finally_after."""

    def before(
        self, hook_context: HookContext, hints: dict[str, Any]
    ) -> EvaluationContext | None:
        LOG.info("Before hook — flag=%s", hook_context.flag_key)
        return None

    def after(
        self,
        hook_context: HookContext,
        details: FlagEvaluationDetails[Any],
        hints: dict[str, Any],
    ) -> None:
        LOG.info("After hook — reason=%s variant=%s", details.reason, details.variant)

    def error(
        self,
        hook_context: HookContext,
        exception: Exception,
        hints: dict[str, Any],
    ) -> None:
        LOG.error("Error hook — flag=%s", hook_context.flag_key, exc_info=exception)

    def finally_after(
        self,
        hook_context: HookContext,
        details: FlagEvaluationDetails[Any],
        hints: dict[str, Any],
    ) -> None:
        LOG.info("Finally After hook — reason=%s", details.reason)
